### PR TITLE
config hierarchy 

### PR DESCRIPTION
**Pull request by LoganWright**
#279

### CODE REVIEW

File: Sources/Vapor/Config/Config.swift [hunk 1]
```diff
@@ -1,258 +1,203 @@
...
-            if files[name] == nil {
-                files[name] = []
-            }
-
-            files[name]?.append(file)
-        }
+    /**
+         Returns whether this instance of `Config` contains the path
+     */
+    public func has(_ path: String, _ indexes: PathIndex...) -> Bool {
```
Comment (LoganWright) [inline Sources/Vapor/Config/Config.swift]: Should we keep this logic? User can check w/ `if let` or `!= nil` themselves. Not sure it's useful

Comment (tannernelson) [inline Sources/Vapor/Config/Config.swift]: I think doing `!=` nil is better. No reason to reinvent operators that Swift already has.

File: Sources/Vapor/Config/Config.swift [hunk 2]
```diff
@@ -1,258 +1,206 @@
...
 
-    /**
-        Returns the generic Json representation for an item at a given path or throws
-     */
-    public func get(_ keyPath: String) throws -> Node {
-        var keys = keyPath.keys
-        guard let json: Json = repository[keys.removeFirst()] else {
-            throw Error.NoFileFound
+        //
+        // TODO: cli will possibly be first in priority queue in future
```
Comment (tannernelson) [inline Sources/Vapor/Config/Config.swift]: Let's finish this TODO before merging.

File: Sources/Vapor/Core/Application.swift [hunk 3]
```diff
@@ -126,29 +101,26 @@ public class Application {
...
 
-        if let handler = self.detectEnvironmentHandler {
-            return handler(environment)
-        } else {
-            return Environment(id: environment)
+        // Tanner, would like input here. 
+        // Alternative is
+        // the json is strictly typed, we need a way to make it fuzzy
+        // or all cli args will be strings.
+        port = config["app", "port"].string?.int ?? 80
```
Comment (LoganWright) [inline Sources/Vapor/Core/Application.swift]: @tannernelson , just flagging this line for you. I think we'll need to make a fuzzier implementation in Json. as the easiest way to handle this.

I have some other ideas, but none that are very good. alternatives appreciated 👍

Comment (tannernelson) [inline Sources/Vapor/Core/Application.swift]: I added issue #293 for this. We can leave the `.string?.int` syntax for now.

File: Sources/Vapor/Config/Config.swift [hunk 4]
```diff
@@ -1,258 +1,280 @@
...
-     */
-    private var repository: [String: Json]
+    private static let suffix = ".json"
 
-    public enum Error: ErrorProtocol {
-        case NoFileFound
-        case NoValueFound
+    init(name: String, json: Json) {
+        if name.hasSuffix(JsonFile.suffix) {
+            let suffixCount = JsonFile.suffix.characters.count
```
Comment (tannernelson) [inline Sources/Vapor/Config/Config.swift]: Could we do `String(name.characters.split(".").first)`?

**Pull request merged**